don't rely on syscall.Kill to interrupt a run / make fortio works on Windows

I just discovered fortio doesn't run on windows:

```
C:\Users\l\go\src\istio.io>go get -u istio.io/fortio
# istio.io/fortio/ui
fortio\ui\uihandler.go:181:3: undefined: syscall.Kill
```

Ignoring for a second that go should provide a kill emulation (imo) - we probably should cleanup fortio anyway as it's only to interrupt itself

(we can also remove the exit option from the UI and only interrupt a given run or maintain a stack or list of them - or interrupt all of them through a shared channel)

--- a/Dockerfile
+++ b/Dockerfile
@@ -9,6 +9,8 @@ COPY . fortio
 RUN CGO_ENABLED=0 GOOS=linux go build -a -ldflags \
   '-X istio.io/fortio/ui.resourcesDir=/usr/local/lib/fortio -X main.defaultDataDir=/var/lib/istio/fortio -s' \
   -o fortio.bin istio.io/fortio
+# Just check it stays compiling on Windows (would need to set the rsrcDir too)
+RUN CGO_ENABLED=0 GOOS=windows go build -a -o fortio.exe istio.io/fortio
 # Minimal image with just the binary and certs
 FROM scratch as release
 COPY --from=build /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/

--- a/periodic/periodic.go
+++ b/periodic/periodic.go
@@ -130,6 +130,11 @@ type periodicRunner struct {
 	RunnerOptions
 }
 
+var (
+	globalAbort chan os.Signal
+	abortMutex  sync.Mutex
+)
+
 // Normalize initializes and normalizes the runner options. In particular it sets
 // up the channel that can be used to interrupt the run later.
 func (r *RunnerOptions) Normalize() {
@@ -163,15 +168,39 @@ func (r *RunnerOptions) Normalize() {
 	}
 	if r.Stop == nil {
 		r.Stop = make(chan struct{}, 1)
-		// Catch SIGINT signals from the OS and close the channel to terminate the run loops
-		c := make(chan os.Signal, 1)
-		signal.Notify(c, os.Interrupt)
 		// TODO: for long running server this slowly leaks go rountines... do better
 		go func() {
-			// in case the signal happens after r.Stop was already closed
-			defer recover() // nolint: errcheck
-			<-c
-			close(r.Stop)
+			abortMutex.Lock()
+			runnerChan := r.Stop
+			if globalAbort == nil {
+				log.Infof("First outstanding run starting, catching signal")
+				globalAbort = make(chan os.Signal, 1)
+				signal.Notify(globalAbort, os.Interrupt)
+			}
+			abortMutex.Unlock()
+			log.LogVf("WATCHER starting new watcher for signal! %d", runtime.NumGoroutine())
+			select {
+			case _, ok := <-globalAbort:
+				log.LogVf("WATCHER Got interrupt signal! %v", ok)
+				abortMutex.Lock()
+				if r.Stop != nil {
+					close(r.Stop)
+					r.Stop = nil
+				}
+				if ok {
+					log.Infof("Closing to notify all and resetting signal handler")
+					if globalAbort != nil {
+						close(globalAbort)
+						globalAbort = nil
+					}
+					signal.Reset(os.Interrupt)
+				}
+				abortMutex.Unlock()
+			case <-runnerChan:
+				log.LogVf("WATCHER r.Stop readable")
+				// nothing to do, stop happened
+			}
+			log.LogVf("WATCHER End of go routine")
 		}()
 	}
 }
@@ -298,6 +327,19 @@ func (r *periodicRunner) Run() RunnerResults {
 	result := RunnerResults{r.Labels, start, requestedQPS, requestedDuration,
 		actualQPS, elapsed, r.NumThreads, Version, functionDuration.Export(r.Percentiles)}
 	result.DurationHistogram.Print(r.Out, "Aggregated Function Time")
+
+	select {
+	case <-r.Stop: // nothing
+		log.LogVf("RUNNER r.Stop already closed")
+	default:
+		log.LogVf("RUNNER r.Stop not already closed, closing")
+		abortMutex.Lock()
+		if r.Stop != nil {
+			close(r.Stop)
+			r.Stop = nil
+		}
+		abortMutex.Unlock()
+	}
 	return result
 }
 
@@ -374,8 +416,6 @@ MainLoop:
 			sleepTimes.Counter.Log(tIDStr + " Sleep time")
 		}
 	}
-	// Put back default handling of ^C (for UI server mode)
-	signal.Reset(os.Interrupt)
 }
 
 func formatDate(d *time.Time) string {

--- a/periodic/periodic.go
+++ b/periodic/periodic.go
@@ -37,7 +37,7 @@ import (
 
 const (
 	// Version is the overall package version (used to version json output too).
-	Version = "0.5.1"
+	Version = "0.5.2"
 )
 
 // DefaultRunnerOptions are the default values for options (do not mutate!).
@@ -86,6 +86,8 @@ type RunnerOptions struct {
 	Out io.Writer
 	// Extra data to be copied back to the results (to be saved/JSON serialized)
 	Labels string
+	// Channel to interrupt a run
+	Stop chan struct{}
 }
 
 // RunnerResults encapsulates the actual QPS observed and duration histogram.
@@ -158,6 +160,18 @@ func newPeriodicRunner(opts *RunnerOptions) *periodicRunner {
 	}
 	if r.Runners == nil {
 		r.Runners = make([]Runnable, r.NumThreads)
+	}
+	if r.Stop == nil {
+		r.Stop = make(chan struct{}, 1)
+		// Catch SIGINT signals from the OS and close the channel to terminate the run loops
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt)
+		// TODO: for long running server this slowly leaks go rountines... do better
+		go func() {
+			defer recover() // in case the signal happens after r.Stop was already closed
+			<-c
+			close(r.Stop)
+		}()
 	}
 	return r
 }
@@ -290,10 +304,6 @@ func runOne(id int, funcTimes *stats.Histogram, sleepTimes *stats.Histogram, num
 	hasDuration := (r.Duration > 0)
 	f := r.Runners[id]
 
-	// Catch SIGINT signals from the OS and raise a flag to terminate the run loop
-	c := make(chan os.Signal, 1)
-	signal.Notify(c, os.Interrupt)
-
 MainLoop:
 	for {
 		fStart := time.Now()
@@ -332,14 +342,14 @@ MainLoop:
 			log.Debugf("%s target next dur %v - sleep %v", tIDStr, targetElapsedDuration, sleepDuration)
 			sleepTimes.Record(sleepDuration.Seconds())
 			select {
-			case <-c:
+			case <-r.Stop:
 				break MainLoop
 			case <-time.After(sleepDuration):
 				// continue normal execution
 			}
 		} else { // Not using QPS
 			select {
-			case <-c:
+			case <-r.Stop:
 				break MainLoop
 			default:
 				// continue to the next iteration

--- a/periodic/periodic.go
+++ b/periodic/periodic.go
@@ -133,9 +133,9 @@ type periodicRunner struct {
 }
 
 var (
-	globalAbort     chan os.Signal
-	outstandingRuns int64
-	abortMutex      sync.Mutex
+	gAbortChan       chan os.Signal
+	gOutstandingRuns int64
+	gAbortMutex      sync.Mutex
 )
 
 // Normalize initializes and normalizes the runner options. In particular it sets
@@ -172,54 +172,57 @@ func (r *RunnerOptions) Normalize() {
 	if r.Stop == nil {
 		r.Stop = make(chan struct{}, 1)
 		go func() {
-			abortMutex.Lock()
-			outstandingRuns++
+			gAbortMutex.Lock()
+			gOutstandingRuns++
 			runnerChan := r.Stop // need a copy to not race with assignement to nil
-			abortChan := globalAbort
-			if globalAbort == nil {
+			abortChan := gAbortChan
+			if gAbortChan == nil {
 				log.Infof("First outstanding run starting, catching signal")
-				globalAbort = make(chan os.Signal, 1)
-				signal.Notify(globalAbort, os.Interrupt)
+				gAbortChan = make(chan os.Signal, 1)
+				signal.Notify(gAbortChan, os.Interrupt)
 			}
-			abortMutex.Unlock()
+			gAbortMutex.Unlock()
 			log.LogVf("WATCHER starting new watcher for signal! %d", runtime.NumGoroutine())
 			select {
 			case _, ok := <-abortChan:
 				log.LogVf("WATCHER Got interrupt signal! %v", ok)
 				r.Abort()
 				if ok {
-					abortMutex.Lock()
-					if globalAbort != nil {
+					gAbortMutex.Lock()
+					if gAbortChan != nil {
 						log.Infof("Closing to notify all")
-						close(globalAbort)
-						globalAbort = nil
+						close(gAbortChan)
+						gAbortChan = nil
 					}
-					abortMutex.Unlock()
+					gAbortMutex.Unlock()
 				}
 			case <-runnerChan:
 				log.LogVf("WATCHER r.Stop readable")
 				// nothing to do, stop happened
 			}
 			log.LogVf("WATCHER End of go routine")
-			abortMutex.Lock()
-			outstandingRuns--
-			if outstandingRuns == 0 {
+			gAbortMutex.Lock()
+			gOutstandingRuns--
+			if gOutstandingRuns == 0 {
 				log.Infof("Last watcher: resetting signal handler")
-				globalAbort = nil
+				gAbortChan = nil
 				signal.Reset(os.Interrupt)
 			}
-			abortMutex.Unlock()
+			gAbortMutex.Unlock()
 		}()
 	}
 }
 
+// Abort safely aborts the run by closing the channel and resetting that channel
+// to nil under lock so it can be called multiple times and not create panic for
+// already closed channel.
 func (r *RunnerOptions) Abort() {
-	abortMutex.Lock()
+	gAbortMutex.Lock()
 	if r.Stop != nil {
 		close(r.Stop)
 		r.Stop = nil
 	}
-	abortMutex.Unlock()
+	gAbortMutex.Unlock()
 }
 
 // internal version, returning the concrete implementation.
@@ -241,9 +244,9 @@ func (r *periodicRunner) Options() *RunnerOptions {
 
 // Run starts the runner.
 func (r *periodicRunner) Run() RunnerResults {
-	abortMutex.Lock()
+	gAbortMutex.Lock()
 	runnerChan := r.Stop // need a copy to not race with assignement to nil
-	abortMutex.Unlock()
+	gAbortMutex.Unlock()
 	useQPS := (r.QPS > 0)
 	hasDuration := (r.Duration > 0)
 	var numCalls int64
@@ -359,7 +362,8 @@ func (r *periodicRunner) Run() RunnerResults {
 }
 
 // runOne runs in 1 go routine.
-func runOne(id int, runnerChan chan struct{}, funcTimes *stats.Histogram, sleepTimes *stats.Histogram, numCalls int64, start time.Time, r *periodicRunner) {
+func runOne(id int, runnerChan chan struct{},
+	funcTimes *stats.Histogram, sleepTimes *stats.Histogram, numCalls int64, start time.Time, r *periodicRunner) {
 	var i int64
 	endTime := start.Add(r.Duration)
 	tIDStr := fmt.Sprintf("T%03d", id)

--- a/periodic/periodic.go
+++ b/periodic/periodic.go
@@ -168,7 +168,6 @@ func (r *RunnerOptions) Normalize() {
 	}
 	if r.Stop == nil {
 		r.Stop = make(chan struct{}, 1)
-		// TODO: for long running server this slowly leaks go rountines... do better
 		go func() {
 			abortMutex.Lock()
 			runnerChan := r.Stop

--- a/periodic/periodic.go
+++ b/periodic/periodic.go
@@ -59,14 +59,14 @@ type Runnable interface {
 
 // MakeRunners creates an array of NumThreads identical Runnable instances.
 // (for the (rare/test) cases where there is no unique state needed)
-func (ro *RunnerOptions) MakeRunners(r Runnable) {
-	log.Infof("Making %d clone of %+v", ro.NumThreads, r)
-	if len(ro.Runners) < ro.NumThreads {
-		log.Infof("Resizing runners from %d to %d", len(ro.Runners), ro.NumThreads)
-		ro.Runners = make([]Runnable, ro.NumThreads)
+func (r *RunnerOptions) MakeRunners(rr Runnable) {
+	log.Infof("Making %d clone of %+v", r.NumThreads, rr)
+	if len(r.Runners) < r.NumThreads {
+		log.Infof("Resizing runners from %d to %d", len(r.Runners), r.NumThreads)
+		r.Runners = make([]Runnable, r.NumThreads)
 	}
-	for i := 0; i < ro.NumThreads; i++ {
-		ro.Runners[i] = r
+	for i := 0; i < r.NumThreads; i++ {
+		r.Runners[i] = rr
 	}
 }
 
@@ -130,9 +130,9 @@ type periodicRunner struct {
 	RunnerOptions
 }
 
-// internal version, returning the concrete implementation.
-func newPeriodicRunner(opts *RunnerOptions) *periodicRunner {
-	r := &periodicRunner{*opts} // by default just copy the input params
+// Normalize initializes and normalizes the runner options. In particular it sets
+// up the channel that can be used to interrupt the run later.
+func (r *RunnerOptions) Normalize() {
 	if r.QPS == 0 {
 		r.QPS = DefaultRunnerOptions.QPS
 	} else if r.QPS < 0 {
@@ -168,11 +168,18 @@ func newPeriodicRunner(opts *RunnerOptions) *periodicRunner {
 		signal.Notify(c, os.Interrupt)
 		// TODO: for long running server this slowly leaks go rountines... do better
 		go func() {
-			defer recover() // in case the signal happens after r.Stop was already closed
+			// in case the signal happens after r.Stop was already closed
+			defer recover() // nolint: errcheck
 			<-c
 			close(r.Stop)
 		}()
 	}
+}
+
+// internal version, returning the concrete implementation.
+func newPeriodicRunner(opts *RunnerOptions) *periodicRunner {
+	r := &periodicRunner{*opts} // by default just copy the input params
+	r.Normalize()
 	return r
 }
 

--- a/periodic/periodic.go
+++ b/periodic/periodic.go
@@ -86,7 +86,9 @@ type RunnerOptions struct {
 	Out io.Writer
 	// Extra data to be copied back to the results (to be saved/JSON serialized)
 	Labels string
-	// Channel to interrupt a run
+	// Channel to interrupt a run. If given non nil, will get closed and nil'ed
+	// by the end of Run(). To abort a run call Runner.Abort(), don't close channel
+	// directly (or risk a panic that it's already closed).
 	Stop chan struct{}
 }
 
@@ -131,8 +133,9 @@ type periodicRunner struct {
 }
 
 var (
-	globalAbort chan os.Signal
-	abortMutex  sync.Mutex
+	globalAbort     chan os.Signal
+	outstandingRuns int64
+	abortMutex      sync.Mutex
 )
 
 // Normalize initializes and normalizes the runner options. In particular it sets
@@ -170,7 +173,9 @@ func (r *RunnerOptions) Normalize() {
 		r.Stop = make(chan struct{}, 1)
 		go func() {
 			abortMutex.Lock()
-			runnerChan := r.Stop
+			outstandingRuns++
+			runnerChan := r.Stop // need a copy to not race with assignement to nil
+			abortChan := globalAbort
 			if globalAbort == nil {
 				log.Infof("First outstanding run starting, catching signal")
 				globalAbort = make(chan os.Signal, 1)
@@ -179,29 +184,42 @@ func (r *RunnerOptions) Normalize() {
 			abortMutex.Unlock()
 			log.LogVf("WATCHER starting new watcher for signal! %d", runtime.NumGoroutine())
 			select {
-			case _, ok := <-globalAbort:
+			case _, ok := <-abortChan:
 				log.LogVf("WATCHER Got interrupt signal! %v", ok)
-				abortMutex.Lock()
-				if r.Stop != nil {
-					close(r.Stop)
-					r.Stop = nil
-				}
+				r.Abort()
 				if ok {
-					log.Infof("Closing to notify all and resetting signal handler")
+					abortMutex.Lock()
 					if globalAbort != nil {
+						log.Infof("Closing to notify all")
 						close(globalAbort)
 						globalAbort = nil
 					}
-					signal.Reset(os.Interrupt)
+					abortMutex.Unlock()
 				}
-				abortMutex.Unlock()
 			case <-runnerChan:
 				log.LogVf("WATCHER r.Stop readable")
 				// nothing to do, stop happened
 			}
 			log.LogVf("WATCHER End of go routine")
+			abortMutex.Lock()
+			outstandingRuns--
+			if outstandingRuns == 0 {
+				log.Infof("Last watcher: resetting signal handler")
+				globalAbort = nil
+				signal.Reset(os.Interrupt)
+			}
+			abortMutex.Unlock()
 		}()
 	}
+}
+
+func (r *RunnerOptions) Abort() {
+	abortMutex.Lock()
+	if r.Stop != nil {
+		close(r.Stop)
+		r.Stop = nil
+	}
+	abortMutex.Unlock()
 }
 
 // internal version, returning the concrete implementation.
@@ -223,6 +241,9 @@ func (r *periodicRunner) Options() *RunnerOptions {
 
 // Run starts the runner.
 func (r *periodicRunner) Run() RunnerResults {
+	abortMutex.Lock()
+	runnerChan := r.Stop // need a copy to not race with assignement to nil
+	abortMutex.Unlock()
 	useQPS := (r.QPS > 0)
 	hasDuration := (r.Duration > 0)
 	var numCalls int64
@@ -281,7 +302,7 @@ func (r *periodicRunner) Run() RunnerResults {
 	sleepTime := stats.NewHistogram(-0.001, 0.001)
 	if r.NumThreads <= 1 {
 		log.Infof("Running single threaded")
-		runOne(0, functionDuration, sleepTime, numCalls, start, r)
+		runOne(0, runnerChan, functionDuration, sleepTime, numCalls, start, r)
 	} else {
 		var wg sync.WaitGroup
 		var fDs []*stats.Histogram
@@ -293,7 +314,7 @@ func (r *periodicRunner) Run() RunnerResults {
 			sDs = append(sDs, sleepP)
 			wg.Add(1)
 			go func(t int, durP *stats.Histogram, sleepP *stats.Histogram) {
-				runOne(t, durP, sleepP, numCalls, start, r)
+				runOne(t, runnerChan, durP, sleepP, numCalls, start, r)
 				wg.Done()
 			}(t, durP, sleepP)
 		}
@@ -328,22 +349,17 @@ func (r *periodicRunner) Run() RunnerResults {
 	result.DurationHistogram.Print(r.Out, "Aggregated Function Time")
 
 	select {
-	case <-r.Stop: // nothing
+	case <-runnerChan: // nothing
 		log.LogVf("RUNNER r.Stop already closed")
 	default:
 		log.LogVf("RUNNER r.Stop not already closed, closing")
-		abortMutex.Lock()
-		if r.Stop != nil {
-			close(r.Stop)
-			r.Stop = nil
-		}
-		abortMutex.Unlock()
+		r.Abort()
 	}
 	return result
 }
 
 // runOne runs in 1 go routine.
-func runOne(id int, funcTimes *stats.Histogram, sleepTimes *stats.Histogram, numCalls int64, start time.Time, r *periodicRunner) {
+func runOne(id int, runnerChan chan struct{}, funcTimes *stats.Histogram, sleepTimes *stats.Histogram, numCalls int64, start time.Time, r *periodicRunner) {
 	var i int64
 	endTime := start.Add(r.Duration)
 	tIDStr := fmt.Sprintf("T%03d", id)
@@ -390,14 +406,14 @@ MainLoop:
 			log.Debugf("%s target next dur %v - sleep %v", tIDStr, targetElapsedDuration, sleepDuration)
 			sleepTimes.Record(sleepDuration.Seconds())
 			select {
-			case <-r.Stop:
+			case <-runnerChan:
 				break MainLoop
 			case <-time.After(sleepDuration):
 				// continue normal execution
 			}
 		} else { // Not using QPS
 			select {
-			case <-r.Stop:
+			case <-runnerChan:
 				break MainLoop
 			default:
 				// continue to the next iteration

--- a/ui/templates/main.html
+++ b/ui/templates/main.html
@@ -15,7 +15,7 @@ Running load test... Results pending...
 <br />
 <progress id="progressBar" max="100" value="0" style="width: 100%"></progress>
 <br />
-<button type="submit" onclick='javascript:fetch("./?exit=Exit");'>Interrupt</button>
+<button type="submit" onclick='javascript:fetch("./?stop=Stop");'>Interrupt</button>
 </div>
 <script>runTestForDuration({{.TestExpectedDurationSeconds}})</script>
 <div class="chart-container" style="position: relative; height:75vh; width:98vw">
@@ -31,8 +31,8 @@ Count axis logarithmic: <input name="ylog" type="checkbox" onclick="updateChart(
 </form>
 </div>
 <pre>{{else}}
-{{if .DoExit}}
-<p>Exiting server as per request.</p>
+{{if .DoStop}}
+<p>Stoping runs as per request.</p>
 {{else}}
 <form>
 <div>
@@ -70,7 +70,7 @@ Debug fetch http://<input type="text" id="uri" name="uri" value="" size=50/>
 <p><i>Or</i></p>
 <form method="POST">
 <div>
-Use with caution, will interrupt/end this server: <input type="submit" name="exit" value="Exit" />
+Interrupt runs from this server: <input type="submit" name="stop" value="Stop" />
 </div>
 </form>
 {{end}}

--- a/ui/templates/main.html
+++ b/ui/templates/main.html
@@ -15,7 +15,7 @@ Running load test... Results pending...
 <br />
 <progress id="progressBar" max="100" value="0" style="width: 100%"></progress>
 <br />
-<button type="submit" onclick='javascript:fetch("./?stop=Stop");'>Interrupt</button>
+<button type="submit" onclick='javascript:fetch("./?stop=Stop&runid={{.RunID}}");'>Interrupt</button>
 </div>
 <script>runTestForDuration({{.TestExpectedDurationSeconds}})</script>
 <div class="chart-container" style="position: relative; height:75vh; width:98vw">

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -58,9 +58,9 @@ var (
 	dataDir        string
 	mainTemplate   *template.Template
 	browseTemplate *template.Template
-	mutex          = &sync.Mutex{}
+	uiRunMapMutex  = &sync.Mutex{}
 	id             int64
-	runs           = make(map[int64]chan struct{})
+	runs           = make(map[int64]*periodic.RunnerOptions)
 )
 
 const (
@@ -169,11 +169,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	}
 	if mode == run {
 		ro.Normalize()
-		mutex.Lock()
+		uiRunMapMutex.Lock()
 		id++ // start at 1 as 0 means interrupt all
 		runid = id
-		runs[runid] = ro.Stop
-		mutex.Unlock()
+		runs[runid] = &ro
+		uiRunMapMutex.Unlock()
 		log.Infof("New run id %d", runid)
 	}
 	if !JSONOnly {
@@ -213,20 +213,20 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	case stop:
 		if runid <= 0 { // Stop all
 			i := 0
-			mutex.Lock()
+			uiRunMapMutex.Lock()
 			for _, v := range runs {
-				close(v)
+				v.Abort()
 				i++
 			}
-			mutex.Unlock()
+			uiRunMapMutex.Unlock()
 			log.Infof("Interrupted %d runs", i)
 		} else { // Stop one
-			mutex.Lock()
-			c := runs[runid]
-			if c != nil {
-				close(c)
+			uiRunMapMutex.Lock()
+			v := runs[runid]
+			if v != nil {
+				v.Abort()
 			}
-			mutex.Unlock()
+			uiRunMapMutex.Unlock()
 		}
 	case run:
 		// mode == run case:
@@ -259,9 +259,9 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			flusher.Flush()
 		}
 		res, err := fhttp.RunHTTPTest(&o)
-		mutex.Lock()
+		uiRunMapMutex.Lock()
 		delete(runs, runid)
-		mutex.Unlock()
+		uiRunMapMutex.Unlock()
 		if err != nil {
 			w.Write([]byte(fmt.Sprintf("Aborting because %s\n", html.EscapeString(err.Error())))) // nolint: errcheck,gas
 			return

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -28,7 +28,7 @@ import (
 	"runtime"
 	"strconv"
 	"strings"
-	"syscall"
+	"sync"
 	"time"
 
 	"istio.io/fortio/fhttp"
@@ -58,6 +58,9 @@ var (
 	dataDir        string
 	mainTemplate   *template.Template
 	browseTemplate *template.Template
+	mutex          = &sync.Mutex{}
+	id             int64
+	runs           = make(map[int64]chan struct{})
 )
 
 const (
@@ -101,10 +104,11 @@ func (w *HTMLEscapeWriter) Write(p []byte) (int, error) {
 // Handler is the main UI handler creating the web forms and processing them.
 func Handler(w http.ResponseWriter, r *http.Request) {
 	LogRequest(r, "UI")
-	DoExit := false
-	if r.FormValue("exit") == "Exit" {
-		log.Critf("Exit request from %v", r.RemoteAddr)
-		DoExit = true
+	DoStop := false
+	runid, _ := strconv.Atoi(r.FormValue("runid")) // nolint: gas
+	if r.FormValue("stop") == "Stop" {
+		log.Critf("Stop request from %v for %d", r.RemoteAddr, runid)
+		DoStop = true
 	}
 	DoLoad := false
 	JSONOnly := false
@@ -158,16 +162,16 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			UpTime                      time.Duration
 			TestExpectedDurationSeconds float64
 			Port                        int
-			DoExit                      bool
+			DoStop                      bool
 			DoLoad                      bool
 		}{r, opts.GetHeaders(), periodic.Version, logoPath, debugPath, chartJSPath,
 			startTime.Format(time.ANSIC), url, labels,
-			fhttp.RoundDuration(time.Since(startTime)), dur.Seconds(), httpPort, DoExit, DoLoad})
+			fhttp.RoundDuration(time.Since(startTime)), dur.Seconds(), httpPort, DoStop, DoLoad})
 		if err != nil {
 			log.Critf("Template execution failed: %v", err)
 		}
 
-		if !DoLoad && !DoExit {
+		if !DoLoad && !DoStop {
 			return
 		}
 
@@ -177,19 +181,15 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		}
 		flusher.Flush()
 	}
-	if DoExit {
-		selfPid := syscall.Getpid()
-		p, err := os.FindProcess(selfPid)
-		if err != nil {
-			log.Critf("Unable to find self process by pid %d: %v", selfPid, err)
-			return
+	if DoStop {
+		i := 0
+		mutex.Lock()
+		for _, v := range runs {
+			close(v)
+			i++
 		}
-		log.Infof("Found process %v for pid %d", p, selfPid)
-		// TODO: Issue #63, this now compiles (unlike syscall.Kill()) but still doesn't work on windows :(
-		err = p.Signal(os.Interrupt)
-		if err != nil {
-			log.Critf("Unable to interrupt self, pid %d: %v", selfPid, err)
-		}
+		mutex.Unlock()
+		log.Infof("Interrupted %d runs", i)
 		return
 	}
 	firstHeader := true
@@ -212,7 +212,7 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	if JSONOnly {
 		out = os.Stderr
 	}
-	ro := periodic.RunnerOptions{
+	ro := periodic.NewPeriodicRunner(&periodic.RunnerOptions{
 		QPS:         qps,
 		Duration:    dur,
 		Out:         out,
@@ -220,14 +220,22 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		Resolution:  resolution,
 		Percentiles: percList,
 		Labels:      labels,
-	}
+	}).Options() // to get the channel initialized - TODO make this more natural
 	o := fhttp.HTTPRunnerOptions{
-		RunnerOptions:      ro,
+		RunnerOptions:      *ro,
 		HTTPOptions:        *opts,
 		AllowInitialErrors: true,
 	}
-
+	mutex.Lock()
+	id++ // start at 1 as 0 means interrupt all
+	thisID := id
+	runs[thisID] = ro.Stop
+	mutex.Unlock()
+	log.LogVf("Run %d", thisID)
 	res, err := fhttp.RunHTTPTest(&o)
+	mutex.Lock()
+	delete(runs, thisID)
+	mutex.Unlock()
 	if err != nil {
 		w.Write([]byte(fmt.Sprintf("Aborting because %s\n", html.EscapeString(err.Error())))) // nolint: errcheck,gas
 		return
@@ -394,7 +402,7 @@ func Serve(port int, debugpath, uipath, staticRsrcDir string, datadir string) {
 	startTime = time.Now()
 	httpPort = port
 	if uipath == "" {
-		fhttp.Serve(port, debugpath) // doesn't return until exit
+		fhttp.Serve(port, debugpath) // doesn't return until stop
 		return
 	}
 	uiPath = uipath

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -218,7 +218,7 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			mutex.Lock()
 			c := runs[runid]
 			if c != nil {
-				close(runs[runid])
+				close(c)
 			}
 			mutex.Unlock()
 		}

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -105,7 +105,7 @@ func (w *HTMLEscapeWriter) Write(p []byte) (int, error) {
 func Handler(w http.ResponseWriter, r *http.Request) {
 	LogRequest(r, "UI")
 	DoStop := false
-	runid, _ := strconv.Atoi(r.FormValue("runid")) // nolint: gas
+	runid, _ := strconv.ParseInt(r.FormValue("runid"), 10, 64) // nolint: gas
 	if r.FormValue("stop") == "Stop" {
 		log.Critf("Stop request from %v for %d", r.RemoteAddr, runid)
 		DoStop = true
@@ -139,8 +139,30 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 	c, _ := strconv.Atoi(r.FormValue("c")) // nolint: gas
-
+	out := io.Writer(os.Stderr)
+	if !JSONOnly {
+		out = io.Writer(&HTMLEscapeWriter{NextWriter: w})
+	}
 	opts := fhttp.NewHTTPOptions(url)
+	ro := periodic.RunnerOptions{
+		QPS:         qps,
+		Duration:    dur,
+		Out:         out,
+		NumThreads:  c,
+		Resolution:  resolution,
+		Percentiles: percList,
+		Labels:      labels,
+	}
+	thisID := int64(0)
+	if DoLoad {
+		ro.Normalize()
+		mutex.Lock()
+		id++ // start at 1 as 0 means interrupt all
+		thisID = id
+		runs[thisID] = ro.Stop
+		mutex.Unlock()
+		log.Infof("New run id %d", thisID)
+	}
 	if !JSONOnly {
 		// Normal html mode
 		if mainTemplate == nil {
@@ -159,13 +181,14 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			StartTime                   string
 			TargetURL                   string
 			Labels                      string
+			RunID                       int64
 			UpTime                      time.Duration
 			TestExpectedDurationSeconds float64
 			Port                        int
 			DoStop                      bool
 			DoLoad                      bool
 		}{r, opts.GetHeaders(), periodic.Version, logoPath, debugPath, chartJSPath,
-			startTime.Format(time.ANSIC), url, labels,
+			startTime.Format(time.ANSIC), url, labels, thisID,
 			fhttp.RoundDuration(time.Since(startTime)), dur.Seconds(), httpPort, DoStop, DoLoad})
 		if err != nil {
 			log.Critf("Template execution failed: %v", err)
@@ -182,16 +205,27 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		flusher.Flush()
 	}
 	if DoStop {
-		i := 0
-		mutex.Lock()
-		for _, v := range runs {
-			close(v)
-			i++
+		if runid <= 0 { // Stop all
+			i := 0
+			mutex.Lock()
+			for _, v := range runs {
+				close(v)
+				i++
+			}
+			mutex.Unlock()
+			log.Infof("Interrupted %d runs", i)
+		} else { // Stop one
+			mutex.Lock()
+			c := runs[runid]
+			if c != nil {
+				close(runs[runid])
+			}
+			mutex.Unlock()
 		}
-		mutex.Unlock()
-		log.Infof("Interrupted %d runs", i)
+
 		return
 	}
+	// DoLoad case:
 	firstHeader := true
 	for _, header := range r.Form["H"] {
 		if len(header) == 0 {
@@ -208,30 +242,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			log.Errf("Error adding custom headers: %v", err)
 		}
 	}
-	out := io.Writer(&HTMLEscapeWriter{NextWriter: w})
-	if JSONOnly {
-		out = os.Stderr
-	}
-	ro := periodic.NewPeriodicRunner(&periodic.RunnerOptions{
-		QPS:         qps,
-		Duration:    dur,
-		Out:         out,
-		NumThreads:  c,
-		Resolution:  resolution,
-		Percentiles: percList,
-		Labels:      labels,
-	}).Options() // to get the channel initialized - TODO make this more natural
 	o := fhttp.HTTPRunnerOptions{
-		RunnerOptions:      *ro,
+		RunnerOptions:      ro,
 		HTTPOptions:        *opts,
 		AllowInitialErrors: true,
 	}
-	mutex.Lock()
-	id++ // start at 1 as 0 means interrupt all
-	thisID := id
-	runs[thisID] = ro.Stop
-	mutex.Unlock()
-	log.LogVf("Run %d", thisID)
 	res, err := fhttp.RunHTTPTest(&o)
 	mutex.Lock()
 	delete(runs, thisID)

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -222,7 +222,6 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			}
 			mutex.Unlock()
 		}
-
 		return
 	}
 	// DoLoad case:
@@ -264,6 +263,9 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		_, err = w.Write(json)
 		if err != nil {
 			log.Errf("Unable to write json output for %v: %v", r.RemoteAddr, err)
+		}
+		if DoSave {
+			SaveJSON(res.ID(), json)
 		}
 		return
 	}

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -185,7 +185,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 		log.Infof("Found process %v for pid %d", p, selfPid)
-		p.Signal(os.Interrupt) //Compiles (unlike syscall.Kill()) but doesn't work on windows :(
+		// TODO: Issue #63, this now compiles (unlike syscall.Kill()) but still doesn't work on windows :(
+		err = p.Signal(os.Interrupt)
+		if err != nil {
+			log.Critf("Unable to interrupt self, pid %d: %v", selfPid, err)
+		}
 		return
 	}
 	firstHeader := true

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -178,7 +178,14 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		flusher.Flush()
 	}
 	if DoExit {
-		syscall.Kill(syscall.Getpid(), syscall.SIGINT) // nolint: errcheck,gas
+		selfPid := syscall.Getpid()
+		p, err := os.FindProcess(selfPid)
+		if err != nil {
+			log.Critf("Unable to find self process by pid %d: %v", selfPid, err)
+			return
+		}
+		log.Infof("Found process %v for pid %d", p, selfPid)
+		p.Signal(os.Interrupt) //Compiles (unlike syscall.Kill()) but doesn't work on windows :(
 		return
 	}
 	firstHeader := true

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -101,28 +101,42 @@ func (w *HTMLEscapeWriter) Write(p []byte) (int, error) {
 
 // TODO: unit tests, allow additional data sets.
 
+type mode int
+
+// The main html has 3 principal modes:
+const (
+	// Default: renders the forms/menus
+	menu mode = iota
+	// Trigger a run
+	run
+	// Request abort
+	stop
+)
+
 // Handler is the main UI handler creating the web forms and processing them.
 func Handler(w http.ResponseWriter, r *http.Request) {
 	LogRequest(r, "UI")
-	DoStop := false
-	runid, _ := strconv.ParseInt(r.FormValue("runid"), 10, 64) // nolint: gas
-	if r.FormValue("stop") == "Stop" {
-		log.Critf("Stop request from %v for %d", r.RemoteAddr, runid)
-		DoStop = true
-	}
-	DoLoad := false
+	mode := menu
 	JSONOnly := false
 	DoSave := (r.FormValue("save") == "on")
 	url := r.FormValue("url")
+	runid := int64(0)
 	if r.FormValue("load") == "Start" {
-		DoLoad = true
+		mode = run
 		if r.FormValue("json") == "on" {
 			JSONOnly = true
 			log.Infof("Starting JSON only load request from %v for %s", r.RemoteAddr, url)
 		} else {
 			log.Infof("Starting load request from %v for %s", r.RemoteAddr, url)
 		}
+	} else {
+		if r.FormValue("stop") == "Stop" {
+			runid, _ = strconv.ParseInt(r.FormValue("runid"), 10, 64) // nolint: gas
+			log.Critf("Stop request from %v for %d", r.RemoteAddr, runid)
+			mode = stop
+		}
 	}
+	// Those only exist/make sense on run mode but go variable declaration...
 	labels := r.FormValue("labels")
 	resolution, _ := strconv.ParseFloat(r.FormValue("r"), 64) // nolint: gas
 	percList, _ := stats.ParsePercentiles(r.FormValue("p"))   // nolint: gas
@@ -134,7 +148,7 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	} else {
 		var err error
 		dur, err = time.ParseDuration(durStr)
-		if DoLoad && err != nil {
+		if mode == run && err != nil {
 			log.Errf("Error parsing duration '%s': %v", durStr, err)
 		}
 	}
@@ -153,15 +167,14 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		Percentiles: percList,
 		Labels:      labels,
 	}
-	thisID := int64(0)
-	if DoLoad {
+	if mode == run {
 		ro.Normalize()
 		mutex.Lock()
 		id++ // start at 1 as 0 means interrupt all
-		thisID = id
-		runs[thisID] = ro.Stop
+		runid = id
+		runs[runid] = ro.Stop
 		mutex.Unlock()
-		log.Infof("New run id %d", thisID)
+		log.Infof("New run id %d", runid)
 	}
 	if !JSONOnly {
 		// Normal html mode
@@ -188,23 +201,16 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			DoStop                      bool
 			DoLoad                      bool
 		}{r, opts.GetHeaders(), periodic.Version, logoPath, debugPath, chartJSPath,
-			startTime.Format(time.ANSIC), url, labels, thisID,
-			fhttp.RoundDuration(time.Since(startTime)), dur.Seconds(), httpPort, DoStop, DoLoad})
+			startTime.Format(time.ANSIC), url, labels, runid,
+			fhttp.RoundDuration(time.Since(startTime)), dur.Seconds(), httpPort, mode == stop, mode == run})
 		if err != nil {
 			log.Critf("Template execution failed: %v", err)
 		}
-
-		if !DoLoad && !DoStop {
-			return
-		}
-
-		flusher, ok := w.(http.Flusher)
-		if !ok {
-			log.Fatalf("expected http.ResponseWriter to be an http.Flusher")
-		}
-		flusher.Flush()
 	}
-	if DoStop {
+	switch mode {
+	case menu:
+		// nothing more to do
+	case stop:
 		if runid <= 0 { // Stop all
 			i := 0
 			mutex.Lock()
@@ -222,67 +228,72 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			}
 			mutex.Unlock()
 		}
-		return
-	}
-	// DoLoad case:
-	firstHeader := true
-	for _, header := range r.Form["H"] {
-		if len(header) == 0 {
-			continue
+	case run:
+		// mode == run case:
+		firstHeader := true
+		for _, header := range r.Form["H"] {
+			if len(header) == 0 {
+				continue
+			}
+			log.LogVf("adding header %v", header)
+			if firstHeader {
+				// If there is at least 1 non empty H passed, reset the header list
+				opts.ResetHeaders()
+				firstHeader = false
+			}
+			err := opts.AddAndValidateExtraHeader(header)
+			if err != nil {
+				log.Errf("Error adding custom headers: %v", err)
+			}
 		}
-		log.LogVf("adding header %v", header)
-		if firstHeader {
-			// If there is at least 1 non empty H passed, reset the header list
-			opts.ResetHeaders()
-			firstHeader = false
+		o := fhttp.HTTPRunnerOptions{
+			RunnerOptions:      ro,
+			HTTPOptions:        *opts,
+			AllowInitialErrors: true,
 		}
-		err := opts.AddAndValidateExtraHeader(header)
+		flusher, ok := w.(http.Flusher)
+		if !ok {
+			log.Fatalf("expected http.ResponseWriter to be an http.Flusher")
+		}
+		if !JSONOnly {
+			flusher.Flush()
+		}
+		res, err := fhttp.RunHTTPTest(&o)
+		mutex.Lock()
+		delete(runs, runid)
+		mutex.Unlock()
 		if err != nil {
-			log.Errf("Error adding custom headers: %v", err)
+			w.Write([]byte(fmt.Sprintf("Aborting because %s\n", html.EscapeString(err.Error())))) // nolint: errcheck,gas
+			return
 		}
-	}
-	o := fhttp.HTTPRunnerOptions{
-		RunnerOptions:      ro,
-		HTTPOptions:        *opts,
-		AllowInitialErrors: true,
-	}
-	res, err := fhttp.RunHTTPTest(&o)
-	mutex.Lock()
-	delete(runs, thisID)
-	mutex.Unlock()
-	if err != nil {
-		w.Write([]byte(fmt.Sprintf("Aborting because %s\n", html.EscapeString(err.Error())))) // nolint: errcheck,gas
-		return
-	}
-	json, err := json.MarshalIndent(res, "", "  ")
-	if err != nil {
-		log.Fatalf("Unable to json serialize result: %v", err)
-	}
-	if JSONOnly {
-		w.Header().Set("Content-Type", "application/json")
-		_, err = w.Write(json)
+		json, err := json.MarshalIndent(res, "", "  ")
 		if err != nil {
-			log.Errf("Unable to write json output for %v: %v", r.RemoteAddr, err)
+			log.Fatalf("Unable to json serialize result: %v", err)
 		}
+		savedAs := ""
 		if DoSave {
-			SaveJSON(res.ID(), json)
+			savedAs = SaveJSON(res.ID(), json)
 		}
-		return
-	}
-	if DoSave {
-		res := SaveJSON(res.ID(), json)
-		if res != "" {
+		if JSONOnly {
+			w.Header().Set("Content-Type", "application/json")
+			_, err = w.Write(json)
+			if err != nil {
+				log.Errf("Unable to write json output for %v: %v", r.RemoteAddr, err)
+			}
+			return
+		}
+		if savedAs != "" {
 			// nolint: errcheck, gas
-			w.Write([]byte(fmt.Sprintf("Saved result to <a href='%s'>%s</a>\n", res, res)))
+			w.Write([]byte(fmt.Sprintf("Saved result to <a href='%s'>%s</a>\n", savedAs, savedAs)))
 		}
+		// nolint: errcheck, gas
+		w.Write([]byte(fmt.Sprintf("All done %d calls %.3f ms avg, %.1f qps\n</pre>\n<script>\n",
+			res.DurationHistogram.Count,
+			1000.*res.DurationHistogram.Avg,
+			res.ActualQPS)))
+		ResultToJsData(w, json)
+		w.Write([]byte("</script></body></html>\n")) // nolint: gas
 	}
-	// nolint: errcheck, gas
-	w.Write([]byte(fmt.Sprintf("All done %d calls %.3f ms avg, %.1f qps\n</pre>\n<script>\n",
-		res.DurationHistogram.Count,
-		1000.*res.DurationHistogram.Avg,
-		res.ActualQPS)))
-	ResultToJsData(w, json)
-	w.Write([]byte("</script></body></html>\n")) // nolint: gas
 }
 
 // ResultToJsData converts a result object to chart data arrays and title

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -222,8 +222,8 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 			log.Infof("Interrupted %d runs", i)
 		} else { // Stop one
 			uiRunMapMutex.Lock()
-			v := runs[runid]
-			if v != nil {
+			v, found := runs[runid]
+			if found {
 				v.Abort()
 			}
 			uiRunMapMutex.Unlock()

--- a/ui/uihandler.go
+++ b/ui/uihandler.go
@@ -417,7 +417,7 @@ func Serve(port int, debugpath, uipath, staticRsrcDir string, datadir string) {
 	startTime = time.Now()
 	httpPort = port
 	if uipath == "" {
-		fhttp.Serve(port, debugpath) // doesn't return until stop
+		fhttp.Serve(port, debugpath) // doesn't return until exit
 		return
 	}
 	uiPath = uipath

